src/snowflake/snowpark/dataframe.py

```diff
@@ -701,7 +706,10 @@ def to_local_iterator(
         )
 
     def __copy__(self) -> "DataFrame":
-        return DataFrame(self._session, copy.copy(self._plan))
+        if isinstance(self._select_statement, MockSelectStatement):
+            return DataFrame(self._session, copy.copy(self._select_statement))
```

Comment:
why copying `self._plan` does not work for mock select statement?

Reply:
Because in `dataframe.py`, the code checks for `self._select_statement` to determine if the SQL simplifier is enabled when this dataframe is constructed. In the current Snowpark Python, copying a dataframe would lose this information and render all following operations non-SQL-simplified. This is a known issue that Yijun and I had tried to tackle, but making `copy` works for SQL simplifier codepath involves changing `expr_to_alias`, which I spent days and gave up on 😢  
In short: `copy` will make a dataframe no longer SQL-simplifiable. This is an existing issue.